clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -422,12 +422,108 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
-static bool isZeroSizedArray(QualType Ty) {
-  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
-    return CATy->getSize() == 0;
+static bool isZeroSizedArray(Sema &SemaRef, QualType Ty) {
+  if (const auto *CAT = SemaRef.getASTContext().getAsConstantArrayType(Ty))
+    return CAT->getSize() == 0;
   return false;
 }
 
+// NOTE: In case this function gets upstreamed and other single source
```

Comment:
I think upstreaming of non-trivial SYCL-related patch is a very long shot that probably may not succeed anyway while our customers will face non-user friendly crashes/problems later in the toolchain, so this should be done as a separate task.
In addition, after Aaron's comments - https://github.com/intel/llvm/pull/4888#discussion_r744929969 , I don't think it is worth it at all.

Reply:
I'm not sure what the required protocol here is. Unless we have have a non-negotiable strict deadline, I believe we should be upstreaming whatever can be upstreamed (even if it delays patch). If the patch in its current form is not up-streamable, that should be fixed. From what I can see we already have required infrastructure for this patch upstream (like `checkTypeSupport`), so I don't see what is blocking an attempt to upstream it. 

@AaronBallman I am not sure I understand your comment here - https://github.com/intel/llvm/pull/4888#discussion_r744929969. This patch is building on top of something which has already been upstreamed. Are you suggesting that will be reverted in the future?

Reply:
Tagging @bader and @zahiraam for their input as well since they drive upstreaming efforts

Reply:
> @AaronBallman I am not sure I understand your comment here - #4888 (comment). This patch is building on top of something which has already been upstreamed. Are you suggesting that will be reverted in the future?

That comment was to suggest that I'm not certain the original code should have landed upstream because it has design issues, and building on top of a questionable foundation is something we want to avoid. However, because it's already landed (for a reasonable amount of time) and downstreams are building on top of it, I think it would cause undue churn to just go and revert it. (That doesn't mean it won't get reverted in the future if someone upstream sees the code and decides it's a problem they won't live with.)

Our goal is to try to upstream everything we can and upstreaming this change fits that goal. To me, the only danger with upstreaming this patch would be that someone in community takes issue with the foundational bits and asks you to rewrite those (which would be a good technical outcome but is not likely planned or particularly trivial effort).

Reply:
So, did I get it right that I should try to upstream this patch?

Reply:
I'm curious what @bader and @zahiraam think, but I would say it's reasonable to upstream the patch (but it could lead to unplanned work).

Reply:
I think whether this patch is up-streamed now or later there will always be un-planed work. It would be my preference to wait for it to be up-streamed only because it was a request from community to do a couple of things before SemaSYCL.cpp gets into an acceptable shape. The handling for special classes (this has already been done) but also the mix if "sema" and "codegen" work that's happening in SemaSYCL.cpp.

Reply:
Just to make sure we are on the same page - this patch actually doesn't depend on things already implemented in SemaSYCL (like kernel generation). I could put the function that this patch adds to SemaSYCL anywhere else, I just did that because it is called only for SYCL. So nothing blocks from upstreaming this patch in parallel with kernel generation.

Reply:
Then upstreaming it is fine. Thanks.

Reply:
Okay then. I've moved it to https://reviews.llvm.org/D114080